### PR TITLE
chore: update eslint to fix linting issues

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

Upgraded eslint to latest and fixed rules along the way.

### CODE REVIEW

File: .babelrc.js [hunk 1]
```diff
@@ -2,9 +2,9 @@ let ignore = [`**/dist`]
 
 // Jest needs to compile this code, but generally we don't want this copied
 // to output folders
-if (process.env.NODE_ENV !== `test`) {
-  ignore.push(`**/__tests__`)
-}
+// if (process.env.NODE_ENV !== `test`) {
+//   ignore.push(`**/__tests__`)
+// }
```
Comment (wardpeet) [inline .babelrc.js:7]: This breaks linting as babel will return an empty ast. Packages should be build with `babel src --out-dir . --ignore \"**/__tests__\"`

File: .eslintrc.js [hunk 2]
```diff
@@ -1,23 +1,22 @@
...
-  parser: "babel-eslint",
+  parser: `@babel/eslint-parser`,
   extends: [
-    "google",
-    "eslint:recommended",
-    "plugin:flowtype/recommended",
-    "plugin:react/recommended",
-    "prettier",
-    "prettier/flowtype",
-    "prettier/react",
```
Comment (wardpeet) [inline .eslintrc.js:12]: these are now inside prettier

File: .eslintrc.js [hunk 3]
```diff
@@ -1,23 +1,22 @@
...
+    `google`,
+    `eslint:recommended`,
+    `plugin:flowtype/recommended`,
+    `plugin:react/recommended`,
+    `prettier`,
   ],
-  plugins: ["flowtype", "prettier", "react", "filenames"],
+  plugins: [`flowtype`, `prettier`, `react`, `filenames`, `@babel`],
```
Comment (wardpeet) [inline .eslintrc.js:10]: Added @babel to augment the this rules

File: .eslintrc.js [hunk 4]
```diff
@@ -39,43 +38,59 @@ module.exports = {
...
+    quotes: [`error`, `backtick`],
+    "react/display-name": `off`,
+    "react/jsx-key": `warn`,
+    "react/no-unescaped-entities": `off`,
+    "react/prop-types": `off`,
+    "require-jsdoc": `off`,
+    "valid-jsdoc": `off`,
+    "prefer-promise-reject-errors": `warn`,
+    "no-prototype-builtins": `warn`,
+    "guard-for-in": `warn`,
```
Comment (wardpeet) [inline .eslintrc.js:78]: These are new rules and moved them to warnings because there were too many.

File: .eslintrc.js [hunk 5]
```diff
@@ -39,43 +38,59 @@ module.exports = {
...
+    "spaced-comment": [`error`, `always`, { markers: [`/`] }],
```
Comment (wardpeet) [inline .eslintrc.js:79]: enable typescript `/// referrence` comments

File: .eslintrc.js [hunk 6]
```diff
@@ -39,43 +38,59 @@ module.exports = {
...
+    camelcase: [
+      `error`,
+      {
+        properties: `never`,
+        ignoreDestructuring: true,
+        allow: [`^unstable_`],
+      },
+    ],
```
Comment (wardpeet) [inline .eslintrc.js:87]: I have no idea why this wasn't necessary before :shrug:

File: packages/gatsby-plugin-netlify-cms/src/cms-identity.js [hunk 7]
```diff
@@ -1,4 +1,4 @@
-/* global __PATH_PREFIX__ CMS_PUBLIC_PATH */
+/* global CMS_PUBLIC_PATH */
```
Comment (wardpeet) [inline packages/gatsby-plugin-netlify-cms/src/cms-identity.js:1]: __PATH_PREFIX__ is in the global globals list

File: packages/gatsby-remark-images-contentful/src/index.js [hunk 8]
```diff
@@ -44,13 +44,7 @@ module.exports = async (
   // This will also allow the use of html image tags
   const rawHtmlNodes = select(markdownAST, `html`)
 
-  const generateImagesAndUpdateNode = async function (node, resolve) {
-    // Ignore if it is not contentful image
-
-    if (node.url.indexOf(`images.ctfassets.net`) === -1) {
-      return resolve()
-    }
```
Comment (wardpeet) [inline packages/gatsby-remark-images-contentful/src/index.js:52]: This is really weird and I'm astonished that this hasn't lead to any bugs.

File: packages/gatsby/src/commands/develop.ts [hunk 9]
```diff
@@ -71,12 +71,12 @@ const doesConfigChangeRequireRestart = (
 const getDebugPort = (port?: number): number => port ?? 9229
 
 export const getDebugInfo = (program: IProgram): IDebugInfo | null => {
-  if (program.hasOwnProperty(`inspect`)) {
+  if (Object.prototype.hasOwnProperty.call(program, `inspect`)) {
```
Comment (wardpeet) [inline packages/gatsby/src/commands/develop.ts:74]: I was fixing this rule but eventually moved to warning.

File: packages/gatsby/src/commands/serve.ts [hunk 10]
```diff
@@ -101,6 +101,7 @@ module.exports = async (program: IServeProgram): Promise<void> => {
   const root = path.join(program.directory, `public`)
 
   const app = express()
+  // eslint-disable-next-lien new-cap
```
Comment (ascorbic) [inline packages/gatsby/src/commands/serve.ts:104]:
```suggestion
  // eslint-disable-next-line new-cap
```

File: packages/gatsby/src/commands/develop.ts (hunk 9 above)
Comment (ascorbic) [reply packages/gatsby/src/commands/develop.ts:74]: Can you explain why this is needed?

File: packages/gatsby/src/utils/webpack-utils.ts [hunk 11]
```diff
@@ -784,7 +784,7 @@ export const createWebpackUtils = (
       ],
       ...eslintConfig(schema, jsxRuntimeExists),
     }
-    //@ts-ignore
+    // @ts-ignore
```
Comment (ascorbic) [inline packages/gatsby/src/utils/webpack-utils.ts:787]: Now I'm wondering why our ts rules are allowing this!

File: packages/gatsby/src/commands/develop.ts (hunk 9 above)
Comment (wardpeet) [reply packages/gatsby/src/commands/develop.ts:74]: https://eslint.org/docs/rules/no-prototype-builtins

It's a security risk as technically program could have overwritten hasOwnProperty

File: packages/gatsby/src/utils/webpack-utils.ts (hunk 11 above)
Comment (wardpeet) [reply packages/gatsby/src/utils/webpack-utils.ts:787]: Do we have a lint rule for the space in typescript?

Comment (ascorbic) [reply packages/gatsby/src/utils/webpack-utils.ts:787]: No, but we do have a rule that requires a description when there's a ts-ignore comment

File: packages/gatsby/src/commands/develop.ts (hunk 9 above)
Comment (ascorbic) [reply packages/gatsby/src/commands/develop.ts:74]: Neat. I'd not thought of that

**Review by ascorbic — DISMISSED** [review]

File: packages/gatsby/src/utils/webpack-utils.ts (hunk 11 above)
Comment (wardpeet) [reply packages/gatsby/src/utils/webpack-utils.ts:787]: It's only a warning :)
```
        "@typescript-eslint/ban-ts-comment": [
          `warn`,
          { "ts-ignore": `allow-with-description` },
        ],
```

**Review by vladar — APPROVED** [review]

**Pull request merged**